ELA/History Default Template

benloh opened the pull request:
This updates the default template for the ELA/History pilot.

The goal is to provide a starting comprehensive set of fields that can be customized for individual projects (by renaming or hiding uneeded fields).

See [VFOI Expansion Template Thinking](https://docs.google.com/spreadsheets/d/1muuzZtvw1yptz9AhF4Lao8eK8VDff_mQdBTF4ZHOFHo/edit?gid=0#gid=0) for a the spec/design.

## To Test

You should test this with a new project template, built from scratch.

1. Check out `dev-bl/new-default`
2. `/.nc.js --dataset=<newproject>`
3. Make sure the necessary fields are there


## Additional Notes
* This includes the necessary changes in `server-template-schema` so that the default template will validate.
* "Force Sort" now has a "No" option to revert to the unsorted types list from the original template file.
* Default nodeDef and edgeDef provenance fields updated.

benloh committed:
warn-replace: Default to "Merge" during Import. Fixes #425

benloh committed:
warn-replace: Add warning about replacing (removing) non-defined edges. Fixes #425.

benloh committed:
new-default: Add new default template that covers both ELA and History projects

Based on https://docs.google.com/spreadsheets/d/1muuzZtvw1yptz9AhF4Lao8eK8VDff_mQdBTF4ZHOFHo/edit?gid=0#gid=0

benloh commented:
@jdanish Please test this to make sure it does what you intend.

jdanish commented:
Can we change the text fields to string instead of markdown?

Also, I just realized - would it be easy to make the default for "force sort" some kind of "no" so it's obvious you can add things in your own sorting order? If not, no big deal.

Can we also add one more othertext on edges? that can be moved to provenance for "Who says this"

benloh committed:
new-default: Default to `strings`, not `markdown`

benloh committed:
new-default: Add `otherText` to edgeDefs as a string.  NOT provenance by default.

benloh commented:
@jdanish Template changes updated.  

NOTE there's a bug if you try to change `isProvenance` to true.  Working on that now, but I might not be able to clear that until tomorrow, depending on the severity.  See #437.

benloh committed:
new-default: Support "Force Sort: No" for color groups. #438

benloh committed:
new-default: Add `othertext` edge def server-template-schema

benloh commented:
@jdanish Please try creating a new template from default and make sure the fields are what you expect.  Also note the unsorted type is now available.  If this looks good we can merge to `deploy-terraform` and test terraforming.

jdanish commented:
I think this works! Some minor things I noted that can be ignored for now:

1. Weight is just there ... I think this is fine but means we would need to hand edit to set any limits. But that's OK. 
2. The order of isProvenance and hide seems to vary ... it'd be nice if it was consistent
3. Some things like node / edge types update immediately, others need a screen reload. Not a big deal but it'd be nice long-term to alert or trigger it if needed. For now, I know it's there so it's fine.

Here is what I made in case that helps ... [StudentConnectionsELA.template.toml.zip](https://github.com/user-attachments/files/21738895/StudentConnectionsELA.template.toml.zip)

jdanish commented:
Also realizing now that I see it, it'd be nice for both nodes and edges to include an InfoSource option that is using InfoOrigin and is in provenance by default. Otherwise the provenance tab has history but no source. Sees redundant, but I'd think we want it to have usable information. Sorry ... I realize I pulled that out but looking through it it feels like a good default if anyone looks.

benloh commented:
> 1. Weight is just there ... I think this is fine but means we would need to hand edit to set any limits. But that's OK.

We do not have a facility for setting limits at the moment.
Only the default value is being set.  But there actually isn't a UI for changing/setting that in the Settings UI currently.  So this is a bigger issue that we'll have to address.  There were a number of ideas about managing template fields that were only ever half implemented.  See #440.  Punting on this for now, but it's a piece of general cleanup/technical debt that we need to address.

> 2. The order of isProvenance and hide seems to vary ... it'd be nice if it was consistent

Good catch.  This is dependent on the order of the `[_ui.*]` spec for the default and current project template.  Will update the default template.

> 3. Some things like node / edge types update immediately, others need a screen reload. Not a big deal but it'd be nice long-term to alert or trigger it if needed. For now, I know it's there so it's fine.

Yeah, there's a fair amount of messaging and lock management cleanup that we still need to do.  It's currently partially documented in #433.  That said, I thought I had caught most of those -- generally I think the values should be updated as soon as you close the Advanced Panel (after clicking "Save Settings").  If there's a specific item that requires a forced reload, please document them in #433.

> 4. Also realizing now that I see it, it'd be nice for both nodes and edges to include an InfoSource option that is using InfoOrigin and is in provenance by default. Otherwise the provenance tab has history but no source. Sees redundant, but I'd think we want it to have usable information. Sorry ... I realize I pulled that out but looking through it it feels like a good default if anyone looks.

Yeah I thought we needed that :).  @jdanish just making sure -- in conversations with Kalani we had decided to call the field `infoOrigin`, but you want the label to be `Info **Source**`, right?  Also, what's the order that you want -- I wasn't sure if I could trust the order of the spreadsheet, since PROVENANCE appears before HISTORY?

For Nodes
```
PROVENANCE
- Citation
- Info Source
---
HISTORY
- Created
- Updated
```

For Edges
```
PROVENANCE
- Citation
- OtherText ("Who says this")
- Info Source
---
HISTORY
- Created
- Updated
```

jdanish commented:
Got it, and all makes sense. Added to 433 (hiding / showing attributes was not updating).

Yes, please use InfoSource (teachers can always relabel).

Re the last one, I updated the spreadsheet to put history after provenance, so it should now match your sample though I moved Other Text after Info Source though I am not sure that matters either way, so I'd say to leave it as-is.

benloh committed:
new-default: Re-add InfoSource back into default template

benloh commented:
@jdanish I've added InfoSource.  Please review one more time to make sure the defaults are what you want.  If so, then I'll merge to `deploy-terraform`.

jdanish commented:
Looks good! Merge away!
